scripts/west_commands/sign.py

```diff
@@ -220,59 +234,118 @@ def sign(self, command, build_dir, bcfg, formats):
+                log.die('imgtool not found; either install it',
+                        '(e.g. "pip3 install imgtool") or provide --tool-path')
+        return tool_path
+
     @staticmethod
     def get_cfg(command, bcfg, item):
         try:
             return bcfg[item]
         except KeyError:
             command.check_force(
-                False,
-                "imgtool parameter unknown, build directory has no {} {}".
-                format('device tree define' if item.startswith('DT_') else
-                       'Kconfig option',
-                       item))
+                False, "build .config is missing a {} value".format(item))
             return None
+
+    @staticmethod
+    def dt_params(b, cache):
+        # Get the flash device's write alignment and the image-0
+        # partition's size out of the build directory's devicetree.
+
+        # Retrieve the list of devicetree bindings from cache.
+        try:
+            bindings = cache.get_list('CACHED_DTS_ROOT_BINDINGS')
+            log.dbg('DTS bindings:', bindings, level=log.VERBOSE_VERY)
+        except KeyError:
+            log.die('CMake cache has no CACHED_DTS_ROOT_BINDINGS.'
+                    '\n  Try again after re-building your application.')
+
+        # Ensure the build directory has a compiled DTS file
+        # where we expect it to be.
+        dts = b / 'zephyr' / (cache['CACHED_BOARD'] + '.dts.pre.tmp')
+        if not dts.is_file():
+            log.die("can't find DTS; expected:", dts)
+        log.dbg('DTS file:', dts, level=log.VERBOSE_VERY)
+
+        # Parse the devicetree using bindings from cache.
+        try:
+            edt = edtlib.EDT(dts, bindings)
+        except edtlib.EDTError as e:
+            log.die("can't parse devicetree:", e)
+
+        # By convention, the zephyr,flash chosen node contains the
+        # partition information about the zephyr image to sign.
+        flash = edt.chosen_node('zephyr,flash')
+        if not flash:
+            log.die('devicetree has no chosen zephyr,flash node;',
+                    "can't infer flash write block or image-0 slot sizes")
+
+        # This node must have a write-block-size property, which
+        # provides the write alignment required by imgtool.
+        if "write-block-size" not in flash.props:
+            log.die('DT zephyr,flash node has no write-block-size;',
+                    "can't determine imgtool write alignment")
+
+        # The node must have a "partitions" child node, which in turn
+        # must have a child node labeled "image-0". By convention, the
+        # primary slot for consumption by imgtool is linked into this
+        # partition.
+        if 'partitions' not in flash.children:
+            log.die("DT zephyr,flash chosen node has no partitions,",
+                    "can't find partition for MCUboot slot 0")
+        partitions = flash.children['partitions']
+        for _, node in partitions.children.items():
```

Comment:
Could use `for node in partitions.children.values()` (and maybe get rid of the `partitions` variable).